decode_payload

```diff
@@ -0,0 +1,130 @@
+      if [ "${CHECKINPUT}" = 1 ] && [ "$(dd status=none bs=1 skip="$((20 + MLEN + OFFSET))" count="${LENGTH}" if="${FILE}" | sha256sum | cut -d ' ' -f 1)" != "$(printf -- "${DATAHASH}" | od -An -vtx1 -w1024 | tr -d ' ')" ]; then
+        echo "Data hash mismatch" >&2
+        exit 1 # Script will continue and fail at the end
+      fi
+    ) &
+    DATAHASH=""
+    RUNNING=$((RUNNING + 1))
+  fi
+  if [ "${RUNNING}" = "${PARALLEL}" ]; then
+    wait
+    RUNNING=0
+  fi
+  if [ "${MODE}" = "new_partition_info" ] && [ "${SIZE}" != 0 ]; then
+    FINALSIZE="${SIZE}"
+  fi
+  if [ "${MODE}" = "new_partition_info" ] && [ "${HASH}" != "" ]; then
+    FINALHASH="${HASH}"
+  fi
+  if [ "${MODE}" = "new_info" ] && [ "${SIZE}" != 0 ]; then
+    KERNELSIZE="${SIZE}"
+  fi
+  if [ "${MODE}" = "new_info" ] && [ "${HASH}" != "" ]; then
+    KERNELHASH="${HASH}"
+  fi
+done <<< "${DESC}"
+wait
+
+if [ "$(stat '-c%s' "${PARTOUT}")" != "${FINALSIZE}" ]; then
+  echo "Size mismatch" >&2
+  exit 1
+fi
+if [ "$(printf -- "${FINALHASH}" | od -An -vtx1 -w1024 | tr -d ' ')" != "$(sha256sum "${PARTOUT}" | cut -d ' ' -f 1)" ]; then
+  echo "Hash mismatch" >&2
+  exit 1
+fi
+if [ "${KERNEL}" != "" ] && [ "$(stat '-c%s' "${KERNEL}")" != "${KERNELSIZE}" ]; then
+  echo "Kernel size mismatch" >&2
+  exit 1
+fi
+if [ "${KERNEL}" != "" ] && [ "$(printf -- "${KERNELHASH}" | od -An -vtx1 -w1024 | tr -d ' ')" != "$(sha256sum "${KERNEL}" | cut -d ' ' -f 1)" ]; then
+  echo "Kernel hash mismatch" >&2
+  exit 1
+fi
+
+SIGDESC=$(protoc --decode=chromeos_update_engine.Signatures --proto_path "${SCRIPTFOLDER}"/src/update_engine "${SCRIPTFOLDER}"/src/update_engine/update_metadata.proto < <(tail -c +$((21 + MLEN + SIGOFFSET)) "${FILE}" |head -c "${SIGSIZE}"))
+echo "${SIGDESC}" >&2
+VERSION=2 # Init for the dev-key case, in the prod-key case the first signature is version 1 and overwrites it, the second is version 2
+while IFS= read -r LINE; do
+  LINE=$(echo "${LINE}" | sed 's/^ *//g')
+  case "${LINE}" in
+  "version:"*) VERSION=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;;
+  "data:"*)
+            SIGDATA=$(echo "${LINE}" | cut -d '"' -f 2- | head -c-2 | sed 's/%/%%/g')
+            # This is a workaround for the dev-key vs prod-key case: sed '/signatures {/d' | sed '/  version: 2/d'
+            SIGHEX=$(printf -- "${SIGDATA}" | sed '/signatures {/d' | sed '/  version: 2/d' | openssl rsautl -verify -pubin -inkey "${PUBKEY}" -raw | tail -c 32 | od -An -vtx1  -w1024 | tr -d ' ')
+            PAYLOADHASH=$(head -c "$((20 + MLEN + SIGOFFSET))" "${FILE}" | sha256sum | cut -d ' ' -f 1)
+            if [ "${VERSION}" = 2 ] && [ "${SIGHEX}" != "${PAYLOADHASH}" ]; then
+              echo "Signature error" >&2
+              exit 1
+            elif [ "${VERSION}" = 1 ]; then
+              echo "Unprocessed version 1 signature: ${SIGHEX} (Payload hash: ${PAYLOADHASH})" >&2
```

Comment:
I need to read the update-engine source code to find out what this version 1 signature is meant to be

Reply:
Found and documented it